src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,465 @@
+                } else {
+                    startId = requestSegments.higherKey(startId);
+                }
+
+                if (startId.equals(SegmentPriority.LOW.name())) {
+                    continue;
+                }
+
+                Segment segment = requestSegments.get(startId);
+                if (segment == null) {
+                    continue;
+                }
+
+                BlockingQueue<RequestType> requests = segment.content;
+
+                if (requests != null && false == requests.isEmpty()) {
+                    clearExpiredRequests(requests);
+
+                    if (false == requests.isEmpty()) {
+                        return Optional.of(requests);
+                    }
+                }
+            }
+
+            Segment segment = requestSegments.get(SegmentPriority.LOW.name());
+
+            if (segment != null) {
+                BlockingQueue<RequestType> requests = segment.content;
+                if (requests != null && false == requests.isEmpty()) {
+                    clearExpiredRequests(requests);
+                    if (false == requests.isEmpty()) {
+                        return Optional.of(requests);
+                    }
+                }
+            }
+            // if we haven't find a non-empty queue , return empty.
+            return Optional.empty();
+        } finally {
+            // it is fine we may have race conditions. We are not trying to
+            // be precise. The objective is to select each segment with equal probability.
+            lastSelectedSegmentId = startId;
+        }
+    }
+
+    private void clearExpiredRequests(BlockingQueue<RequestType> requests) {
+        // In terms of request duration, HCAD throws a request out if it
+        // is older than the detector frequency. This duration limit frees
+        // up HCAD to work on newer requests in the subsequent detection
+        // interval instead of piling up requests that no longer matter.
+        // For example, loading model checkpoints for cache misses requires
+        // a queue configured in front of it. A request contains the checkpoint
+        // document Id and the expiry time, and the queue can hold a considerable
+        // volume of such requests since the size of the request is small.
+        // The expiry time is the start timestamp of the next detector run.
+        // Enforcing the expiry time places an upper bound on each request’s
+        // lifetime.
+        RequestType head = requests.peek();
+        while (head != null && head.getExpirationEpochMs() < clock.millis()) {
+            requests.poll();
+            head = requests.peek();
+        }
+    }
+
+    protected void putOnly(RequestType request) {
+        try {
+            Segment requestQueue = requestSegments
+                .computeIfAbsent(
+                    SegmentPriority.MEDIUM == request.getPriority() ? request.getDetectorId() : request.getPriority().name(),
```

Comment:
Why only consider `MEDIUM` priority here? Can you add some comments?

Reply:
because only medium priority segments use detector id as the key of the segment map.  low and high priority requests just use the segment priority (i.e., low or high) as the key of the segment map.  Added the comments.